### PR TITLE
Fix crazy recursion in VerifyQueueToSeq

### LINKED ISSUE #1652: core_validation: infinite loop on vkWaitForFences validation
Hi,

we encountered a infinite loop on certain circumstances (that are we not sure of yet) if vkWaitForFences  gets validated.
Callstack:
- WaitForFences
- PreCallValidateWaitForFences
- VerifyQueueStateToFence
- VerifyQueueStateToSeq - recursive loop until crash

### PR DESCRIPTION
**Pull request by chrisforbes**
See commit message in second patch for details. Fixes #1652.

### CODE REVIEW

File: layers/core_validation.cpp [hunk 1]
```diff
@@ -4159,38 +4159,58 @@ static void incrementResources(layer_data *dev_data, GLOBAL_CB_NODE *cb_node) {
...
-                for (auto queryEventsPair : cb_node->waitedEventsBeforeQueryReset) {
-                    for (auto event : queryEventsPair.second) {
-                        if (dev_data->eventMap[event].needsSignaled) {
-                            skip |= log_msg(dev_data->report_data, VK_DEBUG_REPORT_ERROR_BIT_EXT,
-                                            VK_DEBUG_REPORT_OBJECT_TYPE_QUERY_POOL_EXT, 0, 0, DRAWSTATE_INVALID_QUERY, "DS",
-                                            "Cannot get query results on queryPool 0x%" PRIx64
-                                            " with index %d which was guarded by unsignaled event 0x%" PRIx64 ".",
-                                            (uint64_t)(queryEventsPair.first.pool), queryEventsPair.first.index, (uint64_t)(event));
+
+    std::unordered_map<QUEUE_STATE *, uint64_t> target_seqs { { initial_queue, initial_seq } };
```
Comment (tobine) [inline layers/core_validation.cpp]: After some squinting and brain pain I think I grok what's going on here but could use some comments to ease learning curve for future generations. Is this the correct understanding?

target_seqs maps queue->sequence that we want to do validate it up to
done_seqs maps queue->sequence that we've completed validation up to

File: layers/core_validation.cpp [hunk 2]
```diff
@@ -4159,38 +4159,58 @@ static void incrementResources(layer_data *dev_data, GLOBAL_CB_NODE *cb_node) {
...
+        for (; seq < target_seq; ++sub_it, ++seq) {
+            for (auto &wait : sub_it->waitSemaphores) {
+                auto other_queue = GetQueueState(dev_data, wait.queue);
+
+                if (other_queue == queue)
+                    continue;   // semaphores /always/ point backwards, so no point here.
+
+                auto other_target_seq = std::max(target_seqs[other_queue], wait.seq);
+                auto other_done_seq = std::max(done_seqs[other_queue], other_queue->seq);
+
```
Comment (tobine) [inline layers/core_validation.cpp]: And here if we're waiting on a separate queue whose wait sequence exceeds what we've checked thus far, we save this new target and add it to worklist for further processing, yes?

File: layers/core_validation.cpp [hunk 3]
```diff
@@ -4159,38 +4159,58 @@ static void incrementResources(layer_data *dev_data, GLOBAL_CB_NODE *cb_node) {
...
+                if (other_done_seq < other_target_seq) {
+                    target_seqs[other_queue] = other_target_seq;
+                    worklist.push_back(other_queue);
+                }
+            }
+
+            for (auto cb : sub_it->cbs) {
```
Comment (tobine) [inline layers/core_validation.cpp]: Now verify for all CB submissions on current queue that all of the waited events have, in fact signaled.

File: layers/core_validation.cpp [hunk 4]
```diff
@@ -4159,38 +4159,58 @@ static void incrementResources(layer_data *dev_data, GLOBAL_CB_NODE *cb_node) {
...
                 }
             }
         }
-        sub_it++;
-        queue_seq++;
-    }
-    for (auto qs : other_queue_seqs) {
-        skip |= VerifyQueueStateToSeq(dev_data, GetQueueState(dev_data, qs.first), qs.second);
+
+        done_seqs[queue] = seq;
```
Comment (tobine) [inline layers/core_validation.cpp]: Finally note that current queue has been validated up to seq.

**Comment by chrisforbes** [comment]
Your understanding is correct. I'll mangle what you said into comments for
V2.

Thanks!

On Apr 6, 2017 6:44 PM, "Tobin Ehlis" <notifications@github.com> wrote:

*@tobine* approved this pull request.

I think this is good but added some comments to make sure I properly
understood the code and would be good to see a few comments in the code to
help future developers.
------------------------------

In layers/core_validation.cpp
<https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/pull/1658#discussion_r110302217>
:

> -            auto &last_seq = other_queue_seqs[wait.queue];
-            last_seq = std::max(last_seq, wait.seq);
-        }
-        for (auto cb : sub_it->cbs) {
-            auto cb_node = GetCBNode(dev_data, cb);
-            if (cb_node) {
-                for (auto queryEventsPair :
cb_node->waitedEventsBeforeQueryReset) {
-                    for (auto event : queryEventsPair.second) {
-                        if (dev_data->eventMap[event].needsSignaled) {
-                            skip |= log_msg(dev_data->report_data,
VK_DEBUG_REPORT_ERROR_BIT_EXT,
-
VK_DEBUG_REPORT_OBJECT_TYPE_QUERY_POOL_EXT, 0, 0,
DRAWSTATE_INVALID_QUERY, "DS",
-                                            "Cannot get query results
on queryPool 0x%" PRIx64
-                                            " with index %d which was
guarded by unsignaled event 0x%" PRIx64 ".",
-
(uint64_t)(queryEventsPair.first.pool), queryEventsPair.first.index,
(uint64_t)(event));
+
+    std::unordered_map<QUEUE_STATE *, uint64_t> target_seqs { {
initial_queue, initial_seq } };

After some squinting and brain pain I think I grok what's going on here but
could use some comments to ease learning curve for future generations. Is
this the correct understanding?

target_seqs maps queue->sequence that we want to do validate it up to
done_seqs maps queue->sequence that we've completed validation up to
------------------------------

In layers/core_validation.cpp
<https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/pull/1658#discussion_r110302453>
:

> +        worklist.pop_back();
+
+        auto target_seq = target_seqs[queue];
+        auto seq = std::max(done_seqs[queue], queue->seq);
+        auto sub_it = queue->submissions.begin() + (seq - queue->seq);
+
+        for (; seq < target_seq; ++sub_it, ++seq) {
+            for (auto &wait : sub_it->waitSemaphores) {
+                auto other_queue = GetQueueState(dev_data, wait.queue);
+
+                if (other_queue == queue)
+                    continue;   // semaphores /always/ point
backwards, so no point here.
+
+                auto other_target_seq =
std::max(target_seqs[other_queue], wait.seq);
+                auto other_done_seq =
std::max(done_seqs[other_queue], other_queue->seq);
+

And here if we're waiting on a separate queue whose wait sequence exceeds
what we've checked thus far, we save this new target and add it to worklist
for further processing, yes?
------------------------------

In layers/core_validation.cpp
<https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/pull/1658#discussion_r110302618>
:

> +            for (auto &wait : sub_it->waitSemaphores) {
+                auto other_queue = GetQueueState(dev_data, wait.queue);
+
+                if (other_queue == queue)
+                    continue;   // semaphores /always/ point
backwards, so no point here.
+
+                auto other_target_seq =
std::max(target_seqs[other_queue], wait.seq);
+                auto other_done_seq =
std::max(done_seqs[other_queue], other_queue->seq);
+
+                if (other_done_seq < other_target_seq) {
+                    target_seqs[other_queue] = other_target_seq;
+                    worklist.push_back(other_queue);
+                }
+            }
+
+            for (auto cb : sub_it->cbs) {

Now verify for all CB submissions on current queue that all of the waited
events have, in fact signaled.
------------------------------

In layers/core_validation.cpp
<https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/pull/1658#discussion_r110302647>
:

>                          }
                     }
                 }
             }
         }
-        sub_it++;
-        queue_seq++;
-    }
-    for (auto qs : other_queue_seqs) {
-        skip |= VerifyQueueStateToSeq(dev_data,
GetQueueState(dev_data, qs.first), qs.second);
+
+        done_seqs[queue] = seq;

Finally note that current queue has been validated up to seq.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/pull/1658#pullrequestreview-31466896>,
or mute the thread

**Comment by tobine** [comment]
Note that AppVeyor build failed although I don't see a log so not clear what happened.

**Comment by chrisforbes** [comment]
Rebased. Just waiting on AppVeyor to decide to even start the build now. Can we get more capacity there?

**Pull request merged**